File: matlab/model/quadcopter.m
function result=quadcopter(motor1,motor2,motor3,motor4)
%QUADCOPTER Summary of this function goes here
%   Detailed explanation goes here

    global thetadot_quad;
    global theta_quad;
    global xdot_quad;
    global x_quad;
    global omega_quad;

    % Physical constants.
    g = 9.81;
    m = 0.5;
    L = 0.25;
    k = 3e-6;
    b = 1e-7;
    I = diag([5e-3, 5e-3, 10e-3]);
    kd = 0.25;
    
    dt = 0.005;

    in = [motor1; motor2; motor3; motor4];

    deviation = 00;
    thetadot_quad = deg2rad(2 * deviation * rand(3, 1) - deviation);
    
    omega_quad = thetadot2omega(thetadot_quad, theta_quad);
    a = acceleration(in, theta_quad, xdot_quad, m, g, k, kd);
    omegadot = angular_acceleration(in, omega_quad, I, L, b, k);

    % Advance system state.
    omega_quad = omega_quad + dt * omegadot;
    thetadot_quad = omega2thetadot(omega_quad, theta_quad); 
    theta_quad = theta_quad + dt * thetadot_quad;
    xdot_quad = xdot_quad + dt * a;
    x_quad = x_quad + dt * xdot_quad;
    
    %x_quad(3) = x_quad(3) -0.11 + 0.2*rand(1);
    
    result = [x_quad;theta_quad];
end


% Compute thrust given current inputs and thrust coefficient.
function T = thrust(inputs, k)
    T = [0; 0; k * sum(inputs)];
end

% Compute torques, given current inputs, length, drag coefficient, and thrust coefficient.
function tau = torques(inputs, L, b, k)
    tau = [
        L * k * (inputs(1) - inputs(3))
        L * k * (inputs(2) - inputs(4))
        b * (inputs(1) - inputs(2) + inputs(3) - inputs(4))
    ];
end

% Compute acceleration in inertial reference frame
% Parameters:
%   g: gravity acceleration
%   m: mass of quadcopter
%   k: thrust coefficient
%   kd: global drag coefficient
function a = acceleration(inputs, angles, vels, m, g, k, kd)
    gravity = [0; 0; -g];
    R = rotation(angles);
    T = R * thrust(inputs, k);
    Fd = -kd * vels;
    a = gravity + 1 / m * T + Fd;
end

% Compute angular acceleration in body frame
% Parameters:
%   I: inertia matrix
function omegad = angular_acceleration(inputs, omega, I, L, b, k)
    tau = torques(inputs, L, b, k);
    omegad = inv(I) * (tau - cross(omega, I * omega));
end

% Convert derivatives of roll, pitch, yaw to omega.
function omega = thetadot2omega(thetadot, angles)
    phi = angles(1);
    theta = angles(2);
    psi = angles(3);
    W = [
        1, 0, -sin(theta)
        0, cos(phi), cos(theta)*sin(phi)
        0, -sin(phi), cos(theta)*cos(phi)
    ];
    omega = W * thetadot;
end

% Convert omega to roll, pitch, yaw derivatives
function thetadot = omega2thetadot(omega, angles)
    phi = angles(1);
    theta = angles(2);
    psi = angles(3);
    W = [
        1, 0, -sin(theta)
        0, cos(phi), cos(theta)*sin(phi)
        0, -sin(phi), cos(theta)*cos(phi)
    ];
    thetadot = inv(W) * omega;
end

File: python/model/quadcopter.py
# Universidade Estadual de Campinas - UNICAMP

# Trabalho Final de Curso
# Diego Pereira Domingos RA 090923
# Guilherme Campos Camargo RA

import numpy as np
from numpy import array,transpose
from math import sin,cos

class quadcopter():

    def __init__(self, model = None):
        self.x = array([.0,.0,.0]); # (x,y,z)
        self.xdot = array([.0,.0,.0]);
        self.theta = array([.0,.0,.0]); # (roll, pitch, yaw)
        self.thetadot = array([.0,.0,.0]);
        self.omega = array([.0,.0,.0]);
        self.omegadot = array([.0,.0,.0]);
        self.model = model
        
    def update(self,dt,inputs):
        self.omega = self.thetadot2omega(self.thetadot, self.theta);
        a = self.acceleration(inputs, self.theta, self.xdot, self.model.m, self.model.g, self.model.k, self.model.kd);
        self.omegadot = self.angular_acceleration(inputs, self.omega, self.model.I, self.model.L, self.model.b, self.model.k);
        self.omega = self.omega + dt * self.omegadot;
        
        self.thetadot = self.omega2thetadot(self.omega, self.theta); 
        self.theta = self.theta + dt * self.thetadot;
        self.xdot = self.xdot + dt * a;
        self.x = self.x + dt * self.xdot;
        
    # Compute thrust given current inputs and thrust coefficient.
    def thrust(self,inputs, k):
        T = array([[0], [0], [k * np.sum([i**2 for i in inputs])]])
        return T
        
    # Compute torques, given current inputs, length, drag coefficient, and thrust coefficient.
    def torques(self,inputs, L, b, k):
        tau = array([.0,.0,.0])
        tau[0] = L*k*(inputs[0]**2-inputs[2]**2)
        tau[1] = L*k*(inputs[1]**2-inputs[3]**2)
        tau[2] = b*(inputs[0]**2-inputs[1]**2+inputs[2]**2-inputs[3]**2)
        return tau
    
    def acceleration(self,inputs, angles, vels, m, g, k, kd):
        gravity = array([0.0, 0.0, -g])
        R = self.rotation(angles);
        T = np.dot(R, self.thrust(inputs, k));
        Ta=np.array([0.0,0.0,0.0])
        Ta[0] = T[0][0]
        Ta[1] = T[1][0]
        Ta[2] = T[2][0]
        Fd = -kd * vels;
        a = gravity + (1. / m) * Ta + Fd;
        return a
    
    def angular_acceleration(self,inputs, omega, I, L, b, k):
        tau = self.torques(inputs, L, b, k)
        omegad = np.dot(np.linalg.inv(I), np.subtract(tau, np.cross(omega, np.dot(I, omega))))      
        # Filtering
        #omegad = self.filterArray(omegad, 1e-14)
        return omegad
        
    # Convert derivatives of roll, pitch, yaw to omega.
    def thetadot2omega(self,thetadot, angles):
        phi_ = angles[0];
        theta_ = angles[1];
        psi_ = angles[2];
        W = array([[1.0, 0.0, -sin(theta_)],[0.0, cos(phi_), cos(theta_)*sin(phi_)],[0.0, -sin(phi_), cos(theta_)*cos(phi_)]]);
        omega = np.dot(W,thetadot);
        return omega    
        
        
    # Convert omega to roll, pitch, yaw derivatives
    def omega2thetadot(self,omega, angles):
        phi_ = angles[0];
        theta_ = angles[1];
        psi_ = angles[2];
        W = array([[1.0, 0.0, -sin(theta_)],[0.0, cos(phi_), cos(theta_)*sin(phi_)],[0.0, -sin(phi_), cos(theta_)*cos(phi_)]]);
        thetadot = np.dot(np.linalg.inv(W), omega);
        return thetadot
    
    
    # Compute rotation matrix for a set of angles.
    def rotation(self,angles):
        phi_ = angles[2] # dont change!!
        theta_ = angles[1] # dont change!!
        psi_ = angles[0] # dont change!!

        R = array([[0.0,0.0,0.0],[0.0,0.0,0.0],[0.0,0.0,0.0]]);
        R[:, 0] = [cos(phi_) * cos(theta_),cos(theta_) * sin(phi_),-sin(theta_)]
        R[:, 1] = [cos(phi_) * sin(theta_) * sin(psi_) - cos(psi_) * sin(phi_),cos(phi_) * cos(psi_) + sin(phi_) * sin(theta_) * sin(psi_),cos(theta_) * sin(psi_)]
        R[:, 2] = [sin(phi_) * sin(psi_) + cos(phi_) * cos(psi_) * sin(theta_),cos(psi_) * sin(phi_) * sin(theta_) - cos(phi_) * sin(psi_),cos(theta_) * cos(psi_)]
        
        return R